add a versionmode aspect to Resources (#332)
xregistry/spec: core/spec.md

specification will be restructured to have clean separation between a
protocol-agnostic core and protocol-specific requirements.

A Registry consists of two main types of entities: Resources and Groups.

Resources typically represent the main data of interest for users of the

> Registry, while Groups,as the name implies, is a mechanism by which related

Author: done

Groups, as the name implies, is a mechanism by which related Resources are
arranged together under a single collection - the Group. The reason for the
grouping is not defined by this specification, so the owners of the Registry
can choose to define (or enforce) any pattern they wish. In this sense, a
Group is similar to a "directory" on a filesystem.

- **Registry**

An implementation of this specification. Typically, the implementation would
include model-specific Groups, Resources and extension attributes.

- **Resource**

Resources represent the main data of interest for the Registry. In the
filesystem analogy, these would be the "files". All Resources exist under a
single Group and, similar to Groups, have a set of Registry metadata.
However, unlike a Group which only has Registry metadata, each Resource can
also have a "document" associated with it. For example, a "schema" Resource
might have a "schema document" as its "document". This specification places no
restriction on the type of content stored in the Resource's document.
Additionally, Resources (unlike Groups) MAY be versioned.

- **Version**

A Version is an instance of a Resource that represents a particular point in

> time state of the Resource. Each Version of a Resource has its own set of

Author: tried to fix it

xRegistry metadata and possibly a domain specific document associated with it.

> Each Resource MUST have at least one Version associated with it.

Reviewer: I think it's worth clarifying here that if versioning isn't important to the client, this one version can remain the only version.

```suggestion
Each Resource MUST have at least one Version associated with it. If versioning is irrelevant for the use case in which the Resource is used, the default Version can be evolved without creating new ones.
```

## Registry Design

As discussed in the [Overview](#overview) section, an xRegistry consists of
two main entities related to the data being managed: Groups and Resources.
However, there are other concepts that make up the overall design and this
section will cover them all in more detail.

**Registry**
An xRegistry instance, or a "Registry", can be thought of as a single rooted
tree of entities as shown in the "xRegistry Core Spec" diagram in the
[Overview](#overview) section. At the root is the Registry entity itself.
This entity is meant to serve a few key purposes:

  - High-level metadata about the Registry itself, such as its creation and
    modified timestamps, name, link to additional documentation, etc.
  - The set of [capabilities](#registry-capabilities) (features) that are

>     supported. For example, does this Registry support filtering of the
>     queries results?

Reviewer: ```suggestion
    supported. For example, does this Registry support filtering of
    query results?
```

Author: done

>   - The domain specific ["model"](#registry-model) that defines the types of

Reviewer: ```suggestion
  - The domain-specific ["model"](#registry-model) that defines the types of
```

    entities being managed by the Registry. For example, the model might
    define a Group called `schemagroups` that has `schemas` as the Resources
    within those Groups.

**Groups**

> Traversing down the tree structure, below the Registry entity there will be

Reviewer: ```suggestion
Traversing down the tree structure, below the Registry entity, there will be
```

Author: done

a set of Groups for the entities managed by the Registry. Each Group is meant
to be a logical grouping of related Resources, much like a directory acts as
a grouping of "files". Groups, like the Registry entity, does have similar
high-level metadata that can be set and can have domain specific extension

> attributes defined.

Author: scanned to whole doc for these

As hinted at with the "directory" analogy, a common use for Groups will be
for them to be light-weight collections without much additional semantics
associated with them. However, this is not a requirement. Because Groups
allow for user-defined extension attributes to be defined, Groups might be
quite rich with respect to managing domain specific data. See the
[Endpoint](../endpoint/spec.md) as an example.

**Resources**
Below, or within, each Group can be a set of Resources. Typically, Resources
are the main pieces of data being managed by the Registry. Like the Registry
and Group entities, that have a set of xRegistry defined "common" metadata

> that can be set, and user defined extension attributes can be defined.

Author: done

actually be accessing one of its Versions. This allows for end-users to
not be concerned with keeping track of which Version is the "default" one as
Versions are added or removed. This is sometimes configured to be the "latest"
Version, however, xRegistry does not mandate that the "default" Version be
the "latest".

While not a requirement, the collection of Versions within a Resource
typically form a directed graph with respect to how they are related.
Meaning, a Version is usually "derived" from another Version known as its
"ancestor". For example, "version 2" might have "version 1" as its
"ancestor". By default, xRegistry assumes that each new Version will have
the current "newest" Version as its ancestor, but this is configurable.
See the [Version Mode](#model.versionmode) section for more information.

**Next Steps**
In summary, the xRegistry design itself is relatively simple and consists
of 4 main concepts to form a tree of entities. However, with these, along
with the extensibility of the xRegistry metadata model, a wide range of
metadata can be categorized, managed and exposed in a consistent way, thus

> allowing for a dynamically discoverable yet interoperable programmatic access

Author: I think one is needed after "interoperable" too. Added both

- Description: The `versionid` of this Version's ancestor.

  The `ancestor` attribute MUST be set to the `versionid` of this Version's
  ancestor. If this Version is a root of an ancestor hierarchy tree then it
  MUST be set to its own `versionid` value.

  See the Resource's [`versionmode`](#model.versionmode) model aspect for

>   more information on how this attribute value can be populate.

Reviewer: ```suggestion
  more information on how this attribute value can be populated.
```

Author: done
